iree/compiler/Codegen/LLVMCPU/LLVMCPUTileAndVectorizeLinalgTensorOps2.cpp

```diff
@@ -0,0 +1,333 @@
+  LogicalResult matchAndRewrite(Operation *op,
+                                PatternRewriter &rewriter) const override {
+    auto contractionOp = dyn_cast<linalg::ContractionOpInterface>(op);
+    if (!contractionOp) return failure();
+
+    linalg::TiledLinalgOp tiledLinalgOp;
+    if (failed(Base::matchAndRewriteBase(op, rewriter, tiledLinalgOp))) {
+      return failure();
+    }
+    rewriter.replaceOp(op, tiledLinalgOp.tensorResults);
+    return success();
+  }
+};
+
+}  // namespace
+
+namespace {
+struct LLVMCPUTileAndVectorizePass2
+    : public LLVMCPUTileAndVectorize2Base<LLVMCPUTileAndVectorizePass2> {
+  LLVMCPUTileAndVectorizePass2(bool vectorize = true)
+      : lowerToVectors(vectorize) {}
+  LLVMCPUTileAndVectorizePass2(const LLVMCPUTileAndVectorizePass2 &pass) {
+    lowerToVectors = pass.lowerToVectors;
+  }
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<linalg::LinalgDialect, memref::MemRefDialect,
+                    vector::VectorDialect>();
+  }
+  void runOnOperation() override;
+
+ private:
+  bool lowerToVectors;
+};
+}  // namespace
+
+void LLVMCPUTileAndVectorizePass2::runOnOperation() {
+  MLIRContext *context = &getContext();
+  auto funcOp = getOperation();
+
+  DEBUG_WITH_TYPE(DEBUG_TYPE, {
+    llvm::dbgs() << "\n--- Before LLVMCPUTileAndVectorizePass2 ---\n";
+    funcOp.print(llvm::dbgs(), OpPrintingFlags().useLocalScope());
+    llvm::dbgs() << "\n\n";
+  });
+
+  // Assume there is a single op with a lowering config we use to drive the
+  // tiling decisions.
+  IREE::Codegen::LoweringConfigAttr config;
+  funcOp.walk([&](linalg::LinalgOp linalgOp) {
+    if (auto opConfig = getLoweringConfig(linalgOp)) {
+      if (opConfig) {
+        // Duplicate configurations.
+        if (config) return signalPassFailure();
+        config = opConfig;
+      }
+    }
+  });
+
+  // Skip vetorization if not all the ops can be vectorized.
+  bool isVectorizable = true;
+  {
+    SmallVector<Operation *> computeOps;
+    SmallVector<LoopTilingAndDistributionInfo> tiledLoops;
+    if (failed(getComputeOps(funcOp, computeOps, tiledLoops))) {
+      return signalPassFailure();
+    }
+
+    for (auto op : computeOps) {
+      if (auto linalgOp = dyn_cast<linalg::LinalgOp>(op)) {
+        if (failed(linalg::vectorizeLinalgOpPrecondition(op))) {
+          isVectorizable = false;
+        }
+      }
+    }
+  }
+
+  // Query if first level of tiling is needed.
+  auto l1TileSizes = config.getTileSizeVals(1);
+  if (!l1TileSizes.empty() && isVectorizable) {
+    // First tile and fuse paralell loops.
+    OpBuilder builder(funcOp.getContext());
+    SmallVector<Operation *> computeOps;
+    SmallVector<LoopTilingAndDistributionInfo> tiledLoops;
+    if (failed(getComputeOps(funcOp, computeOps, tiledLoops))) {
+      return signalPassFailure();
+    }
+    auto consumerOp = dyn_cast<linalg::LinalgOp>(computeOps.back());
+    SmallVector<int64_t> consumerTileSize(
```

Comment:
Ah, this seems odd. It would be better for the `tileConsumerAndFuseProducer` to just use `LinalgTilingOptions`. Could we change that upstream. With that we could avoid getting the `config` value in this pass, but rather have a callback that gets the `config` value by looking at the attribute (see the other places this is done)